Detector/Tracker.hh

```diff
@@ -268,6 +337,114 @@ namespace TrackToy {
+//        std::cout<<"Inner Wall intersection at time: from "<<tinter.begin()<<" to "<<tinter.end()<<std::endl;
+//        std::cout<<"at pos: entering "<<posIn<<" rad "<<posIn.Rho()<<" exiting "<<posEx<<" rad "<<posEx.Rho()<<std::endl;
+
+        simulateElemCross(bfield, mctraj, tinter, xings,inwall_->materialPtr());
+        retval=true;
+      }
+    }
+    return retval;
+  }
+
+  template<class KTRAJ> bool Tracker::ecdwsWallcrossing(KinKal::BFieldMap const& bfield,
+      KinKal::ParticleTrajectory<KTRAJ>& mctraj,
+      std::vector<std::shared_ptr<KinKal::ElementXing<KTRAJ>>>& xings,
+      KinKal::TimeRange& tinter,
+//      std::vector<KinKal::TimeRange>& tinters,
+      double tol
+      ) const {
+    bool retval=false;
+    if (ecpwalldws_!=nullptr) {
+      if (ecpwalldws_->intersect(bfield, mctraj, tinter, tol)) {
+//        auto posIn = mctraj.position3(tinter.begin());
+//        auto posEx = mctraj.position3(tinter.end());
+//        std::cout<<"Endcap Downstream Wall intersection at time: from "<<tinter.begin()<<" to "<<tinter.end()<<std::endl;
+//        std::cout<<"at pos: entering "<<posIn<<" rad "<<posIn.Rho()<<" exiting "<<posEx<<" rad "<<posEx.Rho()<<std::endl;
+
+        simulateElemCross(bfield, mctraj, tinter, xings,ecpwalldws_->materialPtr());
+        retval=true;
+      }
+    }
+    return retval;
+  }
+
+
+  template <class KTRAJ> bool Tracker::simulateElemCross(KinKal::BFieldMap const& bfield, KinKal::ParticleTrajectory<KTRAJ> const& mctraj,
+      KinKal::TimeRange& tinter,
+      std::vector<std::shared_ptr<KinKal::ElementXing<KTRAJ>>>& xings,
+      const MatEnv::DetMaterial* mat) const {
+    using PTCA = KinKal::PiecewiseClosestApproach<KTRAJ,KinKal::Line>;
+    using GELEMXING = KinKal::GenElemXing<KTRAJ>;
+    using GELEMXINGPTR = std::shared_ptr<GELEMXING>;
+    using ROOT::Math::VectorUtil::PerpVector;
+    // create the line representing this hit's element.  The line embeds the timing information
+    auto posIn = mctraj.position3(tinter.begin());
+    auto posOut = mctraj.position3(tinter.end());
+    double path=(posOut-posIn).R();
+
+//    auto posMid = (posOut-posIn)/2;
+//    KinKal::Line const& eline = KinKal::Line(posMid,posMid,tinter.mid(),1.0); //speed set to 1 to avoid division for 0.0
+
+    auto vel = mctraj.velocity(tinter.begin());
+//    KinKal::Line const& eline = KinKal::Line(posIn,tinter.mid(),vel,path);
+    KinKal::VEC3 vprop(0,0,vel.Z());
+    double pathZ=(posOut-posIn).Z();
+    KinKal::Line const& eline = KinKal::Line(posIn,tinter.mid(),vprop,pathZ);
+
+
+    // find the TPOCA between the particle trajectory and the wire line
+//    KinKal::CAHint tphint(tinter.begin(),tinter.mid());
+    KinKal::CAHint tphint(tinter.mid(),tinter.mid());
+    static double tprec(1e-8); // TPOCA precision
+    PTCA tp(mctraj,eline,tphint,tprec);
+    // create the straw xing (regardless of inefficiency)
+    auto xing = std::make_shared<GELEMXING>(tp,mat,path);
```

Comment:
Is this comment old?  Are these GELEMXINGs only for wall material, or do they mix straw and wall xings?

Reply:
yes, the comment was not updated correctly